File: lib/src/main/java/com/github/oliveiradev/lib/OverlapActivity.java
package com.github.oliveiradev.lib;

import android.Manifest;
import android.app.Activity;
import android.content.ClipData;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Binder;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.os.Parcelable;
import android.provider.MediaStore;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.support.v4.app.BundleCompat;

import com.github.oliveiradev.lib.exceptions.CancelOperationException;
import com.github.oliveiradev.lib.exceptions.ExternalStorageWriteException;
import com.github.oliveiradev.lib.exceptions.NotPermissionException;
import com.github.oliveiradev.lib.shared.Constants;
import com.github.oliveiradev.lib.shared.TypeRequest;
import com.github.oliveiradev.lib.util.Utils;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * Created by felipe on 03/05/16.
 */
public class OverlapActivity extends Activity {

    private final static String FILE_URI_EXTRA = "FILE_URI";
    private static final int REQUEST_GALLERY = 1;

    private Uri fileUri;
    private TypeRequest typeRequest;

    @Override
    protected void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState == null) {
            handleIntent(getIntent());
        }
    }

    private Rx2Photo rx2Photo = null;

    private static class Rx2PhotoBinder extends Binder {

        final Rx2Photo rx2Photo;

        private Rx2PhotoBinder(Rx2Photo rx2Photo) {
            this.rx2Photo = rx2Photo;
        }
    }

    public static Intent newIntent(final Context context,  final TypeRequest typeRequest, final Rx2Photo caller) {
        Intent intent = new Intent(context, OverlapActivity.class);
        intent.putExtra(Constants.REQUEST_TYPE_EXTRA, typeRequest);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        final Bundle bundle = new Bundle();
        BundleCompat.putBinder(bundle, Constants.CALLER_EXTRA, new Rx2PhotoBinder(caller));
        intent.putExtra(Constants.CALLER_EXTRA, bundle);
        return intent;
    }

    @Override
    protected void onNewIntent(Intent intent) {
        handleIntent(intent);
    }

    private void handleIntent(Intent intent) {
        typeRequest = (TypeRequest) intent.getExtras().get(Constants.REQUEST_TYPE_EXTRA);
        final Bundle bundle = intent.getExtras().getBundle(Constants.CALLER_EXTRA);
        if (bundle != null) {
            IBinder iBinder = BundleCompat.getBinder(bundle, Constants.CALLER_EXTRA);
            if (iBinder instanceof Rx2PhotoBinder) {
                Rx2PhotoBinder binder = (Rx2PhotoBinder) iBinder;
                rx2Photo = binder.rx2Photo;
            }
        }

        if (hasPermission()) {
            switch (typeRequest) {
                case GALLERY:
                    gallery();
                    break;
                case CAMERA:
                    camera();
                    break;
                case COMBINE:
                    combine(false);
                    break;
                case COMBINE_MULTIPLE:
                    combine(true);
                    break;
            }
        } else {
            requestPermission();
        }
    }

    private void combine(boolean isMultiple) {
        if (!Utils.isExternalStorageWritable()) {
            rx2Photo.propagateThrowable(new ExternalStorageWriteException());
            return;
        }

        fileUri = createImageUri();
        List<Intent> intentList = new ArrayList<>();
        Intent chooserIntent = null;
        Intent pickIntent = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
            pickIntent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, isMultiple);
        }
        intentList = Utils.addIntentsToList(this, intentList, pickIntent);
        Intent takePhotoIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        takePhotoIntent.putExtra(MediaStore.EXTRA_OUTPUT, fileUri);
        intentList = Utils.addIntentsToList(this, intentList, takePhotoIntent);
        if (!intentList.isEmpty()) {
            String title = rx2Photo.getTitle() != null ? rx2Photo.getTitle() : getString(R.string.picker_header);
            chooserIntent = Intent.createChooser(intentList.remove(intentList.size() - 1), title);
            chooserIntent.putExtra(Intent.EXTRA_INITIAL_INTENTS, intentList.toArray(new Parcelable[]{}));
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2 && isMultiple) {
            startActivityForResult(chooserIntent, Constants.REQUEST_CODE_COMBINE_MULPITPLE);
        } else {
            startActivityForResult(chooserIntent, Constants.REQUEST_CODE_COMBINE);
        }
    }

    private void gallery() {
        /*Intent intent = new Intent();
        intent.setType("image*//*");
        intent.setAction(Intent.ACTION_GET_CONTENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);*/
        Intent intent = new Intent(Intent.ACTION_PICK, android.provider.MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        startActivityForResult(intent, Constants.REQUEST_CODE_ATTACH_IMAGE);
    }

    private void camera() {
        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        if (takePictureIntent.resolveActivity(getPackageManager()) != null) {
            if (!Utils.isExternalStorageWritable()) {
                rx2Photo.propagateThrowable(new ExternalStorageWriteException());
                return;
            }
            fileUri = createImageUri();
            takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, fileUri);
            startActivityForResult(takePictureIntent, Constants.REQUEST_CODE_TAKE_PICURE);
        }
    }

    private boolean hasPermission() {
        return Build.VERSION.SDK_INT < Build.VERSION_CODES.M || checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
    }

    private void requestPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_GALLERY);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
            rx2Photo.propagateThrowable(new NotPermissionException(NotPermissionException.RequestEnum.GALLERY));
            finish();
            return;
        }

        switch (typeRequest) {
            case CAMERA:
                camera();
                break;
            case GALLERY:
                gallery();
                break;
            case COMBINE:
                combine(false);
                break;
            case COMBINE_MULTIPLE:
                combine(true);
                break;
        }
    }

    /**
     * If we not choose camera, temp file is unused and must be removed
     */
    private void removeUnusedFile() {
        if (fileUri != null)
            getContentResolver().delete(fileUri, null, null);
    }

    private Uri createImageUri() {
        ContentResolver contentResolver = getContentResolver();
        ContentValues cv = new ContentValues();
        String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(new Date());
        cv.put(MediaStore.Images.Media.TITLE, timeStamp);
        return contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, cv);
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(FILE_URI_EXTRA, fileUri);
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        fileUri = savedInstanceState.getParcelable(FILE_URI_EXTRA);
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == Activity.RESULT_OK) {
            switch (typeRequest) {
                case COMBINE_MULTIPLE:
                    if (data != null && data.getClipData() != null) {
                        ClipData mClipData = data.getClipData();
                        List<Uri> uris = new ArrayList<>();

                        for (int i = 0; i < mClipData.getItemCount(); i++) {
                            uris.add(mClipData.getItemAt(i).getUri());
                        }

                        rx2Photo.onActivityResult(uris);

                        removeUnusedFile();
                    } else if (data != null && data.getData() != null) {
                        List<Uri> uris = new ArrayList<>();
                        uris.add(data.getData());
                        rx2Photo.onActivityResult(uris);

                        removeUnusedFile();
                    } else {
                        rx2Photo.onActivityResult(fileUri);
                    }
                    break;
                case CAMERA:
                    rx2Photo.onActivityResult(fileUri);
                    break;
                case GALLERY:
                case COMBINE:
                    if (data != null && data.getData() != null) {
                        rx2Photo.onActivityResult(data.getData());
                        removeUnusedFile();
                    } else {
                        rx2Photo.onActivityResult(fileUri);
                    }
                    break;
            }
        } else {
            rx2Photo.propagateThrowable(new CancelOperationException(typeRequest));
        }

        finish();
    }
}

File: lib/src/main/java/com/github/oliveiradev/lib/OverlapActivity.kt
package com.github.oliveiradev.lib

import android.Manifest
import android.app.Activity
import android.content.ContentValues
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.*
import android.provider.MediaStore
import android.support.annotation.RequiresApi
import android.support.v4.app.BundleCompat
import com.github.oliveiradev.lib.exceptions.CancelOperationException
import com.github.oliveiradev.lib.exceptions.ExternalStorageWriteException
import com.github.oliveiradev.lib.exceptions.NotPermissionException
import com.github.oliveiradev.lib.shared.Constants
import com.github.oliveiradev.lib.shared.TypeRequest
import com.github.oliveiradev.lib.util.Utils
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by Genius on 03.12.2017.
 */
open class OverlapActivity: Activity() {

    companion object {
        private const val FILE_URI_EXTRA = "FILE_URI"
        private const val REQUEST_GALLERY = 1
        private const val BUNDLE = "bundle"

        fun newIntent(context: Context, typeRequest: TypeRequest, caller: Rx2Photo): Intent {
            val intent = Intent(context, OverlapActivity::class.java)
            intent.putExtra(Constants.REQUEST_TYPE_EXTRA, typeRequest)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            val bundle = Bundle()
            BundleCompat.putBinder(bundle, Constants.CALLER_EXTRA, Rx2PhotoBinder(caller))
            intent.putExtra(Constants.CALLER_EXTRA, bundle)
            return intent
        }
    }

    private var fileUri: Uri? = null
    private lateinit var typeRequest: TypeRequest
    private lateinit var rx2Photo: Rx2Photo

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState == null) {
            handleIntent(intent)
        }
    }

    override fun onNewIntent(intent: Intent) {
        handleIntent(intent)
    }

    private fun handleIntent(intent: Intent) {
        typeRequest = intent.extras.get(Constants.REQUEST_TYPE_EXTRA) as TypeRequest
        val bundle = intent.extras.getBundle(Constants.CALLER_EXTRA)
        if (bundle != null) {
            val iBinder = BundleCompat.getBinder(bundle, Constants.CALLER_EXTRA)
            if (iBinder is Rx2PhotoBinder) {
                rx2Photo = iBinder.rx2Photo
            }
        }

        if (hasPermission()) {
            when (typeRequest) {
                TypeRequest.GALLERY -> gallery()
                TypeRequest.CAMERA -> camera()
                TypeRequest.COMBINE -> combine(false)
                TypeRequest.COMBINE_MULTIPLE -> combine(true)
            }
        } else {
            requestPermission()
        }
    }

    private fun combine(isMultiple: Boolean) {
        if (!Utils.isExternalStorageWritable()) {
            rx2Photo.propagateThrowable(ExternalStorageWriteException())
            return
        }

        fileUri = createImageUri()
        var intentList: MutableList<Intent> = ArrayList()
        var chooserIntent: Intent? = null
        val pickIntent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
            pickIntent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, isMultiple)
        }
        intentList = Utils.addIntentsToList(this, intentList, pickIntent)
        val takePhotoIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        takePhotoIntent.putExtra(MediaStore.EXTRA_OUTPUT, fileUri)
        intentList = Utils.addIntentsToList(this, intentList, takePhotoIntent)
        if (!intentList.isEmpty()) {
            val title = if (rx2Photo.title != null) rx2Photo.title else getString(R.string.picker_header)
            chooserIntent = Intent.createChooser(intentList.removeAt(intentList.size - 1), title)
            chooserIntent.putExtra(Intent.EXTRA_INITIAL_INTENTS, intentList.toTypedArray<Parcelable>())
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2 && isMultiple) {
            startActivityForResult(chooserIntent, Constants.REQUEST_COMBINE_MULTIPLE)
        } else {
            startActivityForResult(chooserIntent, Constants.REQUEST_COMBINE)
        }
    }

    private fun gallery() {
        val intent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
        startActivityForResult(intent, Constants.REQUEST_ATTACH_IMAGE)
    }

    private fun camera() {
        val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        if (takePictureIntent.resolveActivity(packageManager) != null) {
            if (!Utils.isExternalStorageWritable()) {
                rx2Photo.propagateThrowable(ExternalStorageWriteException())
                return
            }
            fileUri = createImageUri()
            takePictureIntent.putExtra(MediaStore.EXTRA_OUTPUT, fileUri)
            startActivityForResult(takePictureIntent, Constants.REQUEST_TAKE_PICTURE)
        }
    }

    private fun hasPermission(): Boolean {
        return Build.VERSION.SDK_INT < Build.VERSION_CODES.M || checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED
    }

    private fun requestPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), REQUEST_GALLERY)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
            rx2Photo.propagateThrowable(NotPermissionException(NotPermissionException.RequestEnum.GALLERY))
            finish()
            return
        }

        when (typeRequest) {
            TypeRequest.CAMERA -> camera()
            TypeRequest.GALLERY -> gallery()
            TypeRequest.COMBINE -> combine(false)
            TypeRequest.COMBINE_MULTIPLE -> combine(true)
        }
    }

    /**
     * If we not choose camera, temp file is unused and must be removed
     */
    private fun removeUnusedFile() {
        if (fileUri != null)
            contentResolver.delete(fileUri, null, null)
    }

    private fun createImageUri(): Uri? {
        val contentResolver = contentResolver
        val cv = ContentValues()
        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
        cv.put(MediaStore.Images.Media.TITLE, timeStamp)
        return contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, cv)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(FILE_URI_EXTRA, fileUri)
        outState.putParcelable(BUNDLE, intent.extras)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        fileUri = savedInstanceState.getParcelable(FILE_URI_EXTRA)

        val bundle = savedInstanceState.getParcelable(BUNDLE) as Bundle
        typeRequest = bundle.get(Constants.REQUEST_TYPE_EXTRA) as TypeRequest
        val caller = bundle.getBundle(Constants.CALLER_EXTRA)
        if (caller != null) {
            val iBinder = BundleCompat.getBinder(caller, Constants.CALLER_EXTRA)
            if (iBinder is Rx2PhotoBinder) {
                rx2Photo = iBinder.rx2Photo
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN)
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (resultCode == Activity.RESULT_OK) {
            when (typeRequest) {
                TypeRequest.COMBINE_MULTIPLE -> if (data != null && data.clipData != null) {
                    val mClipData = data.clipData
                    val uris = (0 until mClipData.itemCount).map { mClipData.getItemAt(it).uri }

                    rx2Photo.onActivityResult(uris)

                    removeUnusedFile()
                } else if (data != null && data.data != null) {
                    val uris = ArrayList<Uri>()
                    uris.add(data.data)
                    rx2Photo.onActivityResult(uris)

                    removeUnusedFile()
                } else if (fileUri != null) {
                    val uris = listOf(fileUri!!)

                    rx2Photo.onActivityResult(uris)
                } else {
                    rx2Photo.onActivityResult(fileUri)
                }
                TypeRequest.CAMERA -> rx2Photo.onActivityResult(fileUri)
                TypeRequest.GALLERY, TypeRequest.COMBINE -> if (data != null && data.data != null) {
                    rx2Photo.onActivityResult(data.data)
                    removeUnusedFile()
                } else {
                    rx2Photo.onActivityResult(fileUri)
                }
            }
        } else {
            rx2Photo.propagateThrowable(CancelOperationException(typeRequest))
            removeUnusedFile()
        }

        finish()
    }

    private class Rx2PhotoBinder constructor(internal val rx2Photo: Rx2Photo) : Binder()
}